scripts/promote.py

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env python
+"""
+A tool for promoting server code to different stages of productions, and minting new releases in github.
+
+In order to release you must have a github access token with permissions to write to your repository. Set
+environment variable `GITHUB_TOKEN_PATH` to the path of a file which contains github access token.
+
+`./promote.py integration` promotes master to integration and creates a prerelease in github.
+`./promote.py staging` promotes integration to staging and creates a prerelease in github.
+`./promote.py production` promotes staging to production and creates a release in github.
+
+Versioning follows https://semver.org/ standard
+"""
+import json
+import tempfile
+
+import requests
+import subprocess
+import argparse
+import os
+import semver
+
+parser = argparse.ArgumentParser(description=__doc__)
+parser.add_argument('stage',
+                    metavar='stage',
+                    type=str,
+                    help="The stage you would like to create a release to.",
+                    choices=["integration", "staging", "production"])
+parser.add_argument('--release', '-r',
+                    type=str,
+                    choices=["major", "minor", "patch", "prerelease", "build"],
+                    default=None,
+                    required=False,
+                    help="The type of release to produce.")
+parser.add_argument('--force', '-f',
+                    action="store_true")
+parser.add_argument('--dry-run', '-d',
+                    action="store_true")
+args = parser.parse_args()
+
+
+def _subprocess(args, **kwargs) -> str:
+    print(f"RUN: {' '.join(args)}")
+    return subprocess.run(args, **kwargs, check=True, stdout=subprocess.PIPE).stdout.decode('utf-8')
+
+
+def check_diff(src, dst):
+    """
+    Check that there are no commits in the src branch that will be overwritten by rebasing on the dst.
+    :param src: the source branch
+    :param dst: the destination branch
+    """
+    result = _subprocess(['git', '--no-pager', 'log', '--graph', '--abbrev-commit', '--pretty=oneline',
+                          '--no-merges', "--", f"{src}", f"^{dst}"])
```

Comment:
This is nit picking, but chopping up bash commands into separate strings for subprocess reduces readability IMO. Consider this:
```
cmd = f"git --no-pager log --graph --abbrev-commit --pretty=oneline --no-merges -- {src} ^{dst}"
_subprocess(cmd.split())
```